shark_turbine/kernel/wave/codegen.py

```diff
@@ -375,31 +399,51 @@ def _construct_gather_scatter_indices(
 
     # As we only support identity input/output mapping for now, we can directly
     # substitute iterators with corresponding expanded index.
-    subs = [(sym, expr.start) for sym, expr in zip(iters.keys(), index.values())]
+    subs = [
+        (sym, expr.start) for sym, expr in zip(iters.keys(), index.values())
+    ] + list(idxc.subs.items())
 
     # Contruct input/output index, substituting iterators in input mapping with
     # expanded index.
     result_index = {key: m.subs(subs) for key, m in zip(symbolc_shape, index_mapping)}
 
     strides = strides_from_symbolic_shape(idxc, symbolc_shape)
     offsets = []
-    subs = [(sym, 0) for sym in iters.keys()]
+
+    start_indices = _get_start_indices(result_index)
+    start_indices_orig = _get_start_indices(index)
+    dynamic_offsets = []
     for i in range(elements_per_thread):
         # Update most-minor dim, i.e. in case of identity mapping it will
         # be equivalent to just vector.load
-        subs[-1] = (subs[-1][0], i)
-        indices = [int(i.subs(subs)) for i in index_mapping]
-        offsets.append(
-            IntegerAttr.get(IndexType.get(), _compute_offset(indices, strides))
+        subs = [(sym, idx) for sym, idx in zip(iters.keys(), start_indices_orig)]
+        subs[-1] = (subs[-1][0], start_indices_orig[-1] + i)
+        indices = [i.subs(subs) for i in index_mapping]
+        offset = _compute_offset(indices, strides) - _compute_offset(
+            start_indices, strides
         )
+        offset = subs_idxc(offset)
+
+        try:
```

Comment:
Instead of the try-except, can we add specific checks that we are looking for? I'm guessing if it is a sympy expression then we dont want to cast to int only when its a sympy integer?

Reply:
Is there any 'canonical' way of checking if sympy expression convertible to int constant?

Reply:
I believe you can use is_number, https://docs.sympy.org/latest/modules/core.html#sympy.core.expr.Expr.is_number

Reply:
done